### PR TITLE
dmenu-rs: add package

### PR DESCRIPTION
**Pull request by benjaminedwardwebb**
dmenu-rs is a pixel perfect port of dmenu, rewritten in Rust with extensive plugin support.

See: https://github.com/Shizcow/dmenu-rs

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by benjaminedwardwebb** [comment]
Ran `nix-env -iA dmenu-rs -f .` on the `nixos-unstable` branch at `52b2ac8ae18bbad4374ff0dd5aeee0fdf1aea739` in my local `nixpkgs` repository on my `aarch64-linux` laptop. Installed successfully and was usable.

**Comment by benjaminedwardwebb** [comment]
Also ran the above with `nix.useSandbox = true;`.

**Comment by benjaminedwardwebb** [comment]
Also, anecdotally, I've been using this (or [a largely similar derivation][1]) as my daily driver launcher for about a week and have observed no issues.

The overlay I use is [here][2].

[2]: https://codeberg.org/benjaminedwardwebb/backpack/src/branch/main/pkgs/dmenu-rs.nix
[1]: https://github.com/benjaminedwardwebb/nixpkgs/tree/dmenu-rs-alternative/pkgs/applications/misc/dmenu-rs

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (benjaminedwardwebb) [inline pkgs/applications/misc/dmenu-rs/default.nix:99]: While I'd like to enable this, the test triggered by `make test` invokes `dmenu` directly.

[It runs][1]

```bash
seq 1 100 | target/dmenu $(ARGS)
```

This not only requires an X11 server but it also hangs waiting for user input.

I was able to hack together the following `checkPhase` that works locally using `timeout`

```bash
timeout --preserve-status 5s xvfb-run make test
```

however simply hoping that 5 seconds is long enough for xvfb-run to spin up and the test to complete on every machine seems like an awful idea. I'm not sure how to kill the hanging call to `dmenu`, no option seems to allow that and I don't know how to feed in a keypress like `<RETURN>` into the `xvfb-run` context.

So I'm just leaving it disabled and relying on the other tests like installing, running the binary, and using it personally on a regular basis.

**Review by benjaminedwardwebb — COMMENTED** [review]
@ElvishJerricco @symphorien thanks for your help in working through some issues that stumped me while working on this derivation.

Of course not a priority, but if either of you have write-access and think the changes are worth merging, I'd appreciate getting this package out of a personal overlay and into `nixpkgs` where others can use it.

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (ElvishJerricco) [inline pkgs/applications/misc/dmenu-rs/default.nix:102]: Why are these needed?

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (ElvishJerricco) [inline pkgs/applications/misc/dmenu-rs/default.nix:68]: Do not use `/build`. It is different depending on the machine. Use `$NIX_BUILD_TOP`.

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (ElvishJerricco) [inline pkgs/applications/misc/dmenu-rs/default.nix:92]: I thought these looked fishy at first, but it actually does make sense. We want the rust platform stuff because it does the vendoring and all that, but it doesn't just do `make` and `make install` like `stdenv.mkDerivation` does by default.

That said, this is very much *not* going to cross compile correctly. Not critical, but unfortunate, and probably fixable.

**Review by ElvishJerricco — CHANGES_REQUESTED** [review]

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (benjaminedwardwebb) [reply pkgs/applications/misc/dmenu-rs/default.nix:95]: I've added a comment that tries to explain these phases' surprising appearance.

I'll be honest, I don't know the exact consequences of this package not cross compiling correctly, but it sure sounds bad. Is it related to not running `./configure` in the `configurePhase`? Should I go read up on Autoconf and figure out how to add that into this derivation?

Comment (ElvishJerricco) [reply pkgs/applications/misc/dmenu-rs/default.nix:95]: I wouldn't worry about it not cross compiling; frankly I don't think you can cross compile a desktop environment with nixpkgs right now anyway. It was just something I found noteworthy. But basically the package's makefile is hostile to cross compilation because e.g. it hardcodes `CC=cc` and doesn't have a way to choose the cargo target. It would need (relatively trivial) changes to their build scripting to fix this.

Comment (benjaminedwardwebb) [reply pkgs/applications/misc/dmenu-rs/default.nix:95]: Thanks for calljng it out. I'm not really a contributor to `dmenu-rs` (I fixed 1 typo) but it's something I'll keep in mind and who knows maybe I can fix it upstream some time in the future.

**Review by ElvishJerricco — APPROVED** [review]
I'm a little unhappy with the `LIBCLANG_PATH` thing; I don't really understand why this package is so picky about using clang and having this lib in this particular environment variable. But it's probably not worth blocking this PR over given that it's probably due to the package's excessively wonky build system.

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (symphorien) [reply pkgs/applications/misc/dmenu-rs/default.nix:104]: to provide clang and the right files to bindgen, use rustPlaftorm.bindgenHook instead.

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (symphorien) [inline pkgs/applications/misc/dmenu-rs/default.nix:73]: why do you need the copy step, and not only the recursive chmod?

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (symphorien) [inline pkgs/applications/misc/dmenu-rs/default.nix:33]: with bindgenHook you presumably need neither libclang, llvm nor clang.

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (symphorien) [reply pkgs/applications/misc/dmenu-rs/default.nix:95]: did you consider using mkDerivation with cargoSetupHook to handle fetching sources?

to quote the manual:
> The `rustPlatform.buildRustPackage` function is split into several hooks: cargoSetupHook to set up vendoring for Cargo-based projects, cargoBuildHook to build a project using Cargo, cargoInstallHook to install a project using Cargo, and cargoCheckHook to run tests in Cargo-based projects. With this change, mixed-language projects can use the relevant hooks within builders other than `buildRustPackage`. However, these changes also required several API changes to `buildRustPackage` itself:

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (symphorien) [inline pkgs/applications/misc/dmenu-rs/default.nix:89]: when you override a full phase, please start with `runHook preBuild` and end with `runHook postBuild`. Same for installPhase.

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (benjaminedwardwebb) [reply pkgs/applications/misc/dmenu-rs/default.nix:73]: `.../cargo-vendor-dir/xcb-0.8.2/` is a symlink to the nix store. I don't think I can `chmod` the nix store inside of the build and in general making anything in the nix store writeable seems conceptually wrong?

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (benjaminedwardwebb) [reply pkgs/applications/misc/dmenu-rs/default.nix:95]: I didn't. That sounds like a more natural way to write this derivation -- as a standard make-based build that happens to include cargo deps. I'll take a stab at rewriting it that way later today. Thanks for this!

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (symphorien) [reply pkgs/applications/misc/dmenu-rs/default.nix:73]: oh it's originally a symlink, I thought it was a copy. My bad.

you can remove the final chmod -R by using cp --no-preserve=mod then

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (benjaminedwardwebb) [reply pkgs/applications/misc/dmenu-rs/default.nix:33]: I was able to drop `llvm` and `libclang`, but it wouldn't build without keeping `clang`.

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (benjaminedwardwebb) [reply pkgs/applications/misc/dmenu-rs/default.nix]: That worked! Thanks for the tip about `rustPlatform.bindgenHook`.

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (symphorien) [reply pkgs/applications/misc/dmenu-rs/default.nix:33]: please drop clang and modify https://github.com/Shizcow/dmenu-rs/blob/master/config.mk#L14 as suggested by the error message

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (symphorien) [inline pkgs/applications/misc/dmenu-rs/default.nix:75]: you can probably run tests with xvfb-run

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (benjaminedwardwebb) [reply pkgs/applications/misc/dmenu-rs/default.nix:33]: Are you suggesting I modify this upstream at `Shizcow/dmenu-rs` or through a patch in the nix derivation?

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (benjaminedwardwebb) [reply pkgs/applications/misc/dmenu-rs/default.nix]: I've updated the `checkPhase` to use the following

```bash
timeout 30s xvfb-run make test || if [ "$?" == "124" ]; then true; else false; fi
```

but ... personally, I would prefer no check phase at all.

This test only passes if the 30 seconds is long enough for the test to run. It seemed to work pretty consistently with 5 seconds on my machine, but using a timeout like this seems destined to cause flaky failures somewhere.

The reason I reached for a timeout was because the `make test` hangs on user input, and I don't how to supply any input keys to the program or avoid this hanging test.

For more detail see: https://github.com/NixOS/nixpkgs/pull/200977#discussion_r1027367488

Comment (symphorien) [reply pkgs/applications/misc/dmenu-rs/default.nix]: My bad I had no read the full discussion. If the tests are not well behaved enough to terminate, let's forget about them.

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (symphorien) [reply pkgs/applications/misc/dmenu-rs/default.nix:34]: downstream is ok (either as a patch or as `substituteInplace config.mk --replace 'clang' '$(CC)'` in postPatch for example)
But hard coding clang as compiler is not distro-friendly in general so you may submit this change upstream if so you wish, after we merge this PR. Using $(CC) is more portable and preferable.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (AndersonTorres) [reply pkgs/applications/misc/dmenu-rs/default.nix:21]: Boy, this is very confusing!

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (AndersonTorres) [inline pkgs/applications/misc/dmenu-rs/default.nix:75]:
```suggestion
```

Mentioning the manual is not needed.

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (AndersonTorres) [inline pkgs/applications/misc/dmenu-rs/default.nix:80]: Too obvious to be commented.

```suggestion
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: maintainers/maintainer-list.nix
Comment (benjaminedwardwebb) [reply maintainers/maintainer-list.nix:1674]: Done :)

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (benjaminedwardwebb) [reply pkgs/applications/misc/dmenu-rs/default.nix:33]: I've added it as a patch here, but also opened [this PR upstream](https://github.com/Shizcow/dmenu-rs/pull/44).

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (benjaminedwardwebb) [reply pkgs/applications/misc/dmenu-rs/default.nix]: Cool, dropped the test.

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (benjaminedwardwebb) [reply pkgs/applications/misc/dmenu-rs/default.nix:20]: Definitely. I started using `buildRustPackage` initially, until @symphorien suggested I use `mkDerivation` and it made the build a lot simpler.

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (benjaminedwardwebb) [reply pkgs/applications/misc/dmenu-rs/default.nix]: Thanks for keeping things concise! :)

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (AndersonTorres) [inline pkgs/applications/misc/dmenu-rs/default.nix:84]: Some more corners to 

1. this should be a postPatch
2. It is better to remove this setting from config.mk:

```suggestion
  postPatch = ''
    sed -i "s@PREFIX = /usr/local@@g" config.mk
  '';
```
3. Use `installFlags = [ "PREFIX=$(out)" ];`.  (Indeed, maybe the previous step becomes unneeded if this flag works.)

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (AndersonTorres) [inline pkgs/applications/misc/dmenu-rs/default.nix:82]: Some more corners to trim

1. this should be a postPatch
2. It is better to remove this setting from config.mk: `sed -i "s@PREFIX = /usr/local@@g" config.mk`
3. Way better: remove this preConfigure completely, and use `installFlags = [ "PREFIX=$(out)" ];` instead.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (benjaminedwardwebb) [reply pkgs/applications/misc/dmenu-rs/default.nix:82]: I actually was trying to go with something like 3 initially, but then decided to just copy what the vanilla C-based `dmenu` was doing in its build: https://github.com/NixOS/nixpkgs/blob/d0a1a0f243138978b1d258056eb5988787da7aac/pkgs/applications/misc/dmenu/default.nix#L21-L23

But I agree, `installFlags` seems more sensible. Let me see if I can get it working.

Comment (benjaminedwardwebb) [reply pkgs/applications/misc/dmenu-rs/default.nix]: Cool, updated to use `installFlags`. I think the first time I was trying to use `makeFlags` and couldn't figure out why it wasn't working -- thanks for suggesting the right way.

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (symphorien) [inline pkgs/applications/misc/dmenu-rs/default.nix:76]: there is no reason to have both postPatch and a patchPhase. As the default patchPhase is still useful if someone downstream wants to add a patch with overrideAttrs, it is probably better to move the content of patchPhase (except runHook) to postPatch.

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (symphorien) [reply pkgs/applications/misc/dmenu-rs/default.nix:33]: I think you forgot to remove the dependency on clang. Thank you for reaching upstream :)

Comment (benjaminedwardwebb) [reply pkgs/applications/misc/dmenu-rs/default.nix]: Oh you're right, woops

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (benjaminedwardwebb) [reply pkgs/applications/misc/dmenu-rs/default.nix]: OK. I've moved everything over into `postPatch`.

But, any guidance on what is a good definition of a patch is / what should generally go in `patchPhase`?

Should it only be applying `.patch` files? Or any change to the project's source code? Any change to the project's build? Any change to the project's dependencies' source code or build?

I guess what this `patchPhase` is currently doing is
  - changing the project's build code (`config.mk` and copying in `Cargo.lock`)
  - subtly change a vendored dependency so that it can be written to, without actually altering any of that dependency's source code or build code

So if drew the line at "patches are things that alter a project's source code (but not necessarily build code)" then it seems like it makes sense to keep it all out of `patchPhase`.

Probably there's no hard rule but just trying to get a feel for what is what in nixpkgs.

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (benjaminedwardwebb) [reply pkgs/applications/misc/dmenu-rs/default.nix]: Removed

File: pkgs/applications/misc/dmenu-rs/default.nix
Comment (AndersonTorres) [reply pkgs/applications/misc/dmenu-rs/default.nix]: > But, any guidance on what is a good definition of a patch is / what should generally go in `patchPhase`?
> 
> Should it only be applying `.patch` files? Or any change to the project's source code? Any change to the project's build? Any change to the project's dependencies' source code or build?

Typically, many distros (if not all!) provide `.patch` files in order to directly modify the source code of the apps they provide. 

A typical case is when a new compiler release does not compile some old code (indeed it happened to me when I tried to compile qemu in Slackware a long time ago...). So, usually a package manager provides patch files for cases like these.
Indeed, many times those patchsets are generated by merely diffing the `master` git branch with the current `stable` git branch, maybe because some people like ultra fresh, pre-alpha code...

Another thing that such package managers can do too is run sanitization scripts that clean up some quirks on the original source code; things like removing annoying lines from a Makefile or fixing hardcoded locations (e.g. FreeBSD ports system installs things on `/usr/local/` instead of `/usr/`; so maybe some tool that expects to find things on /usr will not get it without help).

It gets a bit complicated when both things need to be done: `.patch` files to modify C source code, plus sanitization scripts that cleanup Makefiles.

Now, switch to Nix(pkgs).

In Nixpkgs, we provide both use cases via the same `patchPhase`. However, as I said above, it is a bit complicated...

The typical, unedited patchPhase usually works like this:

```nix
patchPhase = ''
  runHook prePatch

  patch -p1 < <the .patch files provided in patches =[ ... ];>
 
  runHook postPatch
''
```

The `runHook`s are there because of composability:

Nix has the ability of compose overlays of package trees, in a similar manner when you add third party repositories in other package managers (`/etc/apt/sources.lst` anyone?).
Such pre and post hooks allow the creator of an overlay to easily modify the phases without having to rewrite them completely.

Well, when you customize this phase as above, it would look like

```nix
patchPhase = ''
  runHook prePatch

  <a giant script modifying the source code>
 
  runHook postPatch
''
```

However, by doing this, you are basically ignoring any `.patch` files eventually provided in `patches` list.

So, when you need to use such "sanitization scripts", it is better to use `postPatch` instead of rewriting the whole `patchPhase`.

> Probably there's no hard rule but just trying to get a feel for what is what in nixpkgs.

Sometimes many things are developed by trial and error.

Comment (benjaminedwardwebb) [reply pkgs/applications/misc/dmenu-rs/default.nix]: Ah thank you for all that good background knowledge!

**Review by symphorien — APPROVED** [review]

**Comment by symphorien** [comment]
Looks good now! Thank you for bearing with us. One last thing before I merge: this does not build on aarch64 (see https://github.com/NixOS/nixpkgs/pull/200977/checks?check_run_id=10128726771 ). It looks due to upstream but unintended, so please mark it as broken like here

https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/audio/eq10q/default.nix#L36

You can optionally report it upstream, and if they answer that aarch64 support is not in scope for the project, remove the broken attribute and restrict platforms to x86-64-linux only instead.

**Comment by benjaminedwardwebb** [comment]
> Looks good now! Thank you for bearing with us. One last thing before I merge: this does not build on aarch64 (see https://github.com/NixOS/nixpkgs/pull/200977/checks?check_run_id=10128726771 ). It looks due to upstream but unintended, so please mark it as broken like here
> 
> https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/audio/eq10q/default.nix#L36
> 
> You can optionally report it upstream, and if they answer that aarch64 support is not in scope for the project, remove the broken attribute and restrict platforms to x86-64-linux only instead.

@symphorien added `aarch64-linux` as broken. However, I use an `aarch64-linux` machine pretty regularly, so am interested in adding support upstream if possible (and filed [this issue][1] related to it). I don't know how easy this will be, though, so maybe it will have to wait for another PR to nixpkgs.

[1]: https://github.com/Shizcow/dmenu-rs/issues/45

@AndersonTorres Anything holding up an approval?

**Review by AndersonTorres — APPROVED** [review]

**Comment by benjaminedwardwebb** [comment]
Thanks all for the help reviewing and getting this merged! Adding my first package to `nixpkgs` was a very pleasant experience.

**Pull request merged**
